### ISSUE TITLE
[CHECKBOX] "unchecked" behaviour

### ISSUE THREAD

**Issue by nuxodin** [openui/open-ui#313]
I would really like it that when a checkbox is unchecked its value is an empty string.

I don't know about others, but i have to do this hack all the time:
```js
let value = el.value; // other input
if (el.type==='checkbox') { // checkboxes
    value = el.checked ? el.value : '';
}
```

**Comment by nuxodin** [comment]
I also use this "hack" to send an "unchecked" checkbox in a form:

```html
<input type='hidden' value='' name='accepted'>
<input type='checkbox' value='1' name='accepted'>
```

This hack has a rating of 592 at stackoverflow 
https://stackoverflow.com/a/1992745/4865307

And it is mentioned at MDN:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/checkbox

I think if we have the possibilities to redesign the checkbox, we should improve it to this behavior.

```html
<checkbox value="1" name=accepted checked> <!-- checkbox.value === '1' -->
<checkbox value="1" name=accepted> <!-- checkbox.value === ''  -->
```

Conceivable would also be a future expansion with an attribute like "falsy".
```html
<checkbox value="on" falsy="off" name=accepted>
```

**Comment by gregwhitworth** [comment]
@nuxodin thank you so much for filing this. This is the exact same issue that we just discussed in #311 - @domenic actually hits on the confusion of the current checkbox in this [comment](https://github.com/openui/open-ui/issues/311#issuecomment-816248592).

Likewise this issue you raised was brought up to the TAG [here](https://github.com/w3ctag/design-principles/issues/279). Ultimately the group seems to be in agreement that there need to be better cohesion in how the initial states relate to that of the IDL values themselves.

As far as your proposals are concerned they need to follow the proposal for `open` in `<popup>` in that we should be clear in what is controlling the initial state vs reflection in the IDL. Effectively the groups seem to be leaning towards making it clear in the name that you're setting the initial state & value rather than that content attribute reflecting the value. @domenic do you want to add any additional thoughts here?

**Comment by nuxodin** [comment]
My final solution would be:

```html
<checkbox on="yes" off="no" name=accepted>
```
```js
checkbox.value // no
checkbox.checked = true;
checkbox.value // yes
```

Advantages: 
- Also unchecked checkboxes sends something (as default an empty string)
- the property "value" no longer incorrectly reflects the attribute "value" (the attribute no longer exists)
- "on" and "off" are short and clear attributes

**Comment by nuxodin** [comment]
This suggestion was discussed and it was concluded that traditional check boxes should also be given an off value (as I understand it)
https://github.com/whatwg/html/issues/4180#issuecomment-1645108696

This will probably look something like this:

```html
<input type=checkbox value=yes name=accepted **offvalue=no**>.
```

Unfortunately this will result in `checkbox.value` not being able to have the selected value (like checkbox.value), but always the on-state value, as this needs to remain compatible :(

So we still have to do something like this:

```js
let value = el.value; // other input
if (el.type==='checkbox') { // checkboxes
    // value = el.checked ? el.value : ''; // old
    value = el.checked ? el.value : el.offValue;
}
```
I would rather there were new checkboxes that had a better API.

**Comment by annevk** [comment]
I don't think that's necessarily true @nuxodin. With an opt-in attribute we could make `el.value` take that into account. The `value` IDL attribute and content attribute are different after all. `defaultValue` is the one that reflects the `value` content attribute.

**Comment by nuxodin** [comment]
Do you mean something like that?

Current behavior

```html
<input type=checkbox value="true" checked> 
// .value == "true"
<input type=checkbox value="val">
// .value == "true" // i think .value should be "" (empty string)
```

Future

```html
<input type=checkbox value="true" offvalue="false" checked>
// .value == "true"
<input type=checkbox value="true" offvalue="false">
// .value == "false"
```

**Comment by annevk** [comment]
Yes, something like that. What would help motivate such a feature would be more evidence that this is a problem web developers frequently run into. Stack Overflow questions might help with that as well as JS libraries that work around the issue.

**Comment by lukewarlow** [comment]
As far as evidence the Vue.js framework has a special API specifically for checkboxes.

They have a true-value and false-value binding. https://vuejs.org/guide/essentials/forms#checkbox-1

Something like a new uncheckedvalue attribute set to false would seem to be in a similar space.

Fwiw this is something I commonly see, people often want checkboxes (and switches) to be a Boolean value rather than submitting conditionally.

**Comment by keithamus** [comment]
I need to WHATWG this in some way.

**Comment by lukewarlow** [comment]
Just went to open a new issue but see this one exists so we'll keep this one open. This came up as a passing comment on my review of #1470 I think this is something that would be good to address. The behaviour of an unchecked checkbox (and now switch) not submitting any value has just never sat right with me. I think a new attribute like `uncheckedvalue="some other value"` (name to be bikeshedded to oblivion I'm sure), that changes the submission behaviour and then also making it so that .value reflects this value when unchecked would be great.

I guess a further question I have is what about indeterminate. Do we need a third value available for that state? How does indeterminate work today regarding form submission? I know indeterminate isn't mutually exclusive with checked/unchecked (as much as css might try to say otherwise).

**Comment by frehner** [comment]
>How does indeterminate work today regarding form submission? I know indeterminate isn't mutually exclusive with checked/unchecked (as much as css might try to say otherwise).

from https://html.spec.whatwg.org/multipage/input.html#the-input-element

> Each [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element has an indeterminateness, a boolean which is initially false. It only has an effect on [checkbox](https://html.spec.whatwg.org/multipage/input.html#checkbox-state-(type=checkbox)) controls, including their rendering and accessibility semantics. An [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element's indeterminateness is independent of its [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked).

So it doesn't affect the form submission in any way is my understanding.

Relatedly, it would be nice to add an attribute for indeterminate -- it's annoying that it's only a property. :)

**Comment by lukewarlow** [comment]
Yeah an `initiallyindeterminate` attribute (name tbd) would be neat. I'm guessing an issue for that already exists in whatwg/html, but perhaps if not worth opening one here or there.

**Comment by frehner** [comment]
This is tracking the attribute proposal I believe. https://github.com/whatwg/html/issues/6578
